packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts

```diff
@@ -0,0 +1,252 @@
+
+  readonly listStackResources: ListStackResources;
+}
+
+export class EvaluateCloudFormationTemplate {
+  private readonly stackResources: ListStackResources;
+  private readonly context: { [k: string]: string };
+  private readonly account: string;
+  private readonly region: string;
+  private readonly partition: string;
+
+  constructor(props: EvaluateCloudFormationTemplateProps) {
+    this.stackResources = props.listStackResources;
+    this.context = {
+      'AWS::AccountId': props.account,
+      'AWS::Region': props.region,
+      'AWS::Partition': props.partition,
+      'AWS::URLSuffix': props.urlSuffix,
+      ...props.parameters,
+    };
+    this.account = props.account;
+    this.region = props.region;
+    this.partition = props.partition;
+  }
+
+  public async findPhysicalNameFor(logicalId: string): Promise<string | undefined> {
+    const stackResources = await this.stackResources.listStackResources();
+    return stackResources.find(sr => sr.LogicalResourceId === logicalId)?.PhysicalResourceId;
+  }
+
+  public async evaluateCfnExpression(cfnExpression: any): Promise<any> {
+    const self = this;
+    class CfnIntrinsics {
+      public evaluateIntrinsic(intrinsic: Intrinsic): any {
+        const intrinsicFunc = (this as any)[intrinsic.name];
+        if (!intrinsicFunc) {
+          throw new CfnEvaluationException(`CloudFormation function ${intrinsic.name} is not supported`);
+        }
+
+        const argsAsArray = Array.isArray(intrinsic.args) ? intrinsic.args : [intrinsic.args];
+
+        return intrinsicFunc.apply(this, argsAsArray);
+      }
+
+      async 'Fn::Join'(separator: string, args: any[]): Promise<string> {
+        const evaluatedArgs = await self.evaluateCfnExpression(args);
+        return evaluatedArgs.join(separator);
+      }
+
+      async 'Fn::Split'(separator: string, args: any): Promise<string> {
+        const evaluatedArgs = await self.evaluateCfnExpression(args);
+        return evaluatedArgs.split(separator);
+      }
+
+      async 'Fn::Select'(index: number, args: any[]): Promise<string> {
+        const evaluatedArgs = await self.evaluateCfnExpression(args);
+        return evaluatedArgs[index];
+      }
+
+      async 'Ref'(logicalId: string): Promise<string> {
+        const refTarget = await self.findRefTarget(logicalId);
+        if (refTarget) {
+          return refTarget;
+        } else {
+          throw new CfnEvaluationException(`Parameter or resource '${logicalId}' could not be found for evaluation`);
+        }
+      }
+
+      async 'Fn::GetAtt'(logicalId: string, attributeName: string): Promise<string> {
+        // ToDo handle the 'logicalId.attributeName' form of Fn::GetAtt
+        const attrValue = await self.findGetAttTarget(logicalId, attributeName);
+        if (attrValue) {
+          return attrValue;
+        } else {
+          throw new CfnEvaluationException(`Attribute '${attributeName}' of resource '${logicalId}' could not be found for evaluation`);
```

Comment:
This error is somewhat conflating the cases: 

1. No resource with that logical ID
2. We have no idea how to get that particular attribute from this particular type of resource

One is a mistake by the user, the other one is a deficiency in the system that needs to be fixed. Would be great to be able to tell them apart.

Why not throw the error at the site of `return undefined` ?

Reply:
Yep, you are right. Corrected.